File: backend/src/main/kotlin/model/Faction.kt
package net.slingray.model

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.descriptors.element
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonPrimitive
import net.slingray.model.Expansion.BASE
import net.slingray.model.Expansion.POK

@Serializable(with = FactionSerializer::class)
enum class Faction(val long: String, val short: String, val expansion: Expansion) : Comparable<Faction> {
    ARBOREC("The Arborec", "arborec", BASE),
    ARGENT("The Argent Flight", "argent", POK),
    BARONY("The Barony of Letnev", "barony", BASE),
    SAAR("The Clan of Saar", "saar", BASE),
    MUAAT("The Embers of Muaat", "muaat", BASE),
    HACAN("The Emirates of Hacan", "hacan", BASE),
    EMPYREAN("The Empyrean", "empyrean", POK),
    SOL("The Federation of Sol", "sol", BASE),
    GHOSTS("The Ghosts of Creuss", "ghosts", BASE),
    L1Z1X("The L1Z1X Mindnet", "l1z1x", BASE),
    MAHACT("The Mahact Gene-Sorcerers", "mahact", POK),
    MENTAK("The Mentak Coalition", "mentak", BASE),
    NAALU("The Naalu Collective", "naalu", BASE),
    NAAZ_ROKHA("The Naaz-Rokha Alliance", "naaz-rokha", POK),
    NEKRO("The Nekro Virus", "nekro", BASE),
    NOMAD("The Nomad", "nomad", POK),
    SARDAKK("Sardakk N'orr", "sardakk", BASE),
    TITANS("The Titans of Ul", "titans", POK),
    JOL_NAR("The Universities of Jol-Nar", "jol-nar", BASE),
    VUIL_RAITH("The Vuil'Raith Cabal", "vuil'raith", POK),
    WINNU("The Winnu", "winnu", BASE),
    XXCHA("The Xxcha Kingdom", "xxcha", BASE),
    YIN("The Yin Brotherhood", "yin", BASE),
    YSSARIL("The Yssaril Tribes", "yssaril", BASE)
}

val factionComparator = Comparator<Faction> { f1, f2 ->
    f1.long.removePrefix("The ").compareTo(
        f2.long.removePrefix("The ")
    )
}

object FactionSerializer : KSerializer<Faction> {
    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("Faction") {
        element<String>("long")
        element<String>("short")
        element<Expansion>("expansion")
    }

    override fun serialize(encoder: Encoder, value: Faction) {
        val composite = encoder.beginStructure(descriptor)
        composite.encodeStringElement(descriptor, 0, value.long)
        composite.encodeStringElement(descriptor, 1, value.short)
        composite.encodeSerializableElement(descriptor, 2, ExpansionSerializer, value.expansion)
        composite.endStructure(descriptor)
    }

    override fun deserialize(decoder: Decoder): Faction {
        if (decoder is JsonDecoder) {
            val element = decoder.decodeJsonElement()
            return when (element) {
                is JsonPrimitive -> {
                    val value = element.content
                    Faction.entries.firstOrNull { it.long == value || it.short == value }
                        ?: throw IllegalArgumentException("Unknown faction: $value")
                }

                is JsonObject -> {
                    val longValue = element["long"]?.jsonPrimitive?.content
                    val shortValue = element["short"]?.jsonPrimitive?.content
                    Faction.entries.firstOrNull { it.long == longValue || it.short == shortValue }
                        ?: throw IllegalArgumentException("Unknown faction: $longValue")
                }

                else -> throw IllegalArgumentException("Invalid JSON for Faction: $element")
            }
        } else {
            throw IllegalStateException("FactionSerializer only works with JSON")
        }
    }
}

File: src/types/faction.ts
import {EXPANSIONS} from "./expansion.ts";

export const FACTIONS = {
    ARBOREC: { long: "The Arborec", short: "arborec", expansion: EXPANSIONS.BASE },
    ARGENT: { long: "The Argent Flight", short: "argent", expansion: EXPANSIONS.POK },
    BARONY: { long: "The Barony of Letnev", short: "barony", expansion: EXPANSIONS.BASE },
    SAAR: { long: "The Clan of Saar", short: "saar", expansion: EXPANSIONS.BASE },
    COUNCIL: { long: "The Council Keleres", short: "council", expansion: EXPANSIONS.TE },
    CRIMSON: { long: "The Crimson Rebellion", short: "crimson", expansion: EXPANSIONS.TE },
    DEEPWROUGHT: { long: "The Deepwrought Scholarate", short: "deepwrought", expansion: EXPANSIONS.TE },
    MUAAT: { long: "The Embers of Muaat", short: "muaat", expansion: EXPANSIONS.BASE },
    HACAN: { long: "The Emirates of Hacan", short: "hacan", expansion: EXPANSIONS.BASE },
    EMPYREAN: { long: "The Empyrean", short: "empyrean", expansion: EXPANSIONS.POK },
    SOL: { long: "The Federation of Sol", short: "sol", expansion: EXPANSIONS.BASE },
    FIRM_OBS: { long: "The Firmament / The Obsidian", short: "firm-obs", expansion: EXPANSIONS.TE },
    FIRMAMENT: { long: "The Firmament", short: "firmament", expansion: EXPANSIONS.TE },
    FRACTURE: { long: "The Fracture", short: "fracture", expansion: EXPANSIONS.TE },
    GHOSTS: { long: "The Ghosts of Creuss", short: "ghosts", expansion: EXPANSIONS.BASE },
    L1Z1X: { long: "The L1Z1X Mindnet", short: "l1z1x", expansion: EXPANSIONS.BASE },
    BASTION: { long: "Last Bastion", short: "bastion", expansion: EXPANSIONS.TE },
    MAHACT: { long: "The Mahact Gene-Sorcerers", short: "mahact", expansion: EXPANSIONS.POK },
    MENTAK: { long: "The Mentak Coalition", short: "mentak", expansion: EXPANSIONS.BASE },
    NAALU: { long: "The Naalu Collective", short: "naalu", expansion: EXPANSIONS.BASE },
    NAAZ_ROKHA: { long: "The Naaz-Rokha Alliance", short: "naaz-rokha", expansion: EXPANSIONS.POK },
    NEKRO: { long: "The Nekro Virus", short: "nekro", expansion: EXPANSIONS.BASE },
    NOMAD: { long: "The Nomad", short: "nomad", expansion: EXPANSIONS.POK },
    OBSIDIAN: { long: "The Obsidian", short: "obsidian", expansion: EXPANSIONS.TE },
    RAL_NEL: { long: "The Ral Nel Consortium", short: "ral-nel", expansion: EXPANSIONS.TE },
    SARDAKK: { long: "Sardakk N'orr", short: "sardakk", expansion: EXPANSIONS.BASE },
    TITANS: { long: "The Titans of Ul", short: "titans", expansion: EXPANSIONS.POK },
    JOL_NAR: { long: "The Universities of Jol-Nar", short: "jol-nar", expansion: EXPANSIONS.BASE },
    VUIL_RAITH: { long: "The Vuil'Raith Cabal", short: "vuil'raith", expansion: EXPANSIONS.POK },
    WINNU: { long: "The Winnu", short: "winnu", expansion: EXPANSIONS.BASE },
    XXCHA: { long: "The Xxcha Kingdom", short: "xxcha", expansion: EXPANSIONS.BASE },
    YIN: { long: "The Yin Brotherhood", short: "yin", expansion: EXPANSIONS.BASE },
    YSSARIL: { long: "The Yssaril Tribes", short: "yssaril", expansion: EXPANSIONS.BASE },
}

export type Faction = typeof FACTIONS[keyof typeof FACTIONS];

export function compareFactions(a: Faction, b: Faction) {
    const normalize = (s: string) =>
        s.startsWith("The ") ? s.slice(4) : s

    return normalize(a.long).localeCompare(normalize(b.long))
}